design/authentication-flow-usability.md

```diff
@@ -0,0 +1,322 @@
+    required-actions and consent screens. This may not work for custom authentication pages added by user themselves, in case
+    they don't use the generic Keycloak freemarker template. We can't do much here and JS won't be triggered in that case. We
+    just need to document this in the server developer guide and encourage users that it is recommended to inherit their
+    authenticators from the common Keycloak freemarker template. Also we need to doublecheck that our
+    authenticator/requiredActions quickstarts work with that.
+    
+    * In case that user triggers new authentication request in browser tab2 when there is already authentication in progress
+    in tab1, KC will be able to lookup current authenticationSession and directly redirect user to the latest step. So for
+    example assuming that user is already on the TOTP screen in tab1 and he starts new authentication request to OIDC endpoint
+    in tab2, then he will be redirected to TOTP screen in tab2.
+    
+    * This also means that after authentication completed in tab1, KC redirects to the client application in particular tab as in
+    current Keycloak master. But in other browser tabs, there won't be immediate redirect to the applications in that case as this can have
+    some side-effects. For example unecessarily creating the HTTP session for the application where user was redirected in other browser tabs.
+    So in other browser tabs, we may display some "info" stating like "Authentication was completed. Redirecting to the
+    application in few seconds". There will be also some additional JS code, which will be triggered when user enters the particular browser tab. This JS
+    will refresh the page and will automatically redirect user to the application.
+
+
+3. As mentioned above, the client data, which are specific to each browser tab, won't be saved in the authentication session
+anymore, but will be saved in the browser itself. This will be either in the request parameter sent across authentication requests or in the cookie, which will need to be "namespaced" to be specific per each browser tab. Few details to this:
+    
+    * The client data will be encoded in the JWS token. It can be pretty similar to the current KC_RESTART cookie - in other words, some JSON data signed with the realm secret key.
+    
+    * In ideal case, the token can be added as separate parameter to the authentication requests. In other words, instead of
+    parameters `code`, `execution`, `client_data` and `tab_id`, we will have `code`, `execution`, `tab_id` and `client_data_full` where
+    `client_data_full` will contain this encoded token. No need for `client_data` parameter, which has just few of client data.
+    
+    * There is a URL limit of 2000 characters (excluding host, but including the path, query params and fragment). So to be
+    on safe side, we can't include `client_data_full` as request parameter in case that is too long. I suggest to use 1200
```

Comment:
There is also limit on cookie size, number of cookies per domain, and total size of all cookies in a domain which varies across the browsers. This limit can be only 4096 bytes for the total size per http://browsercookielimits.squawky.net/. With all the info stored in the auth session notes, and some 10+ number of tabs, this limit could be reached quickly.

Since the limit is based on cookie domain and not cookie path, I don't think the boilerplate suffix would help here.

We thus need to find another way. Could we use POST instead of GET to pass the state? Perhaps part of the state could be stored in the URL (likely the `code`, `execution`, `tab_id`, `client_data` from minimal approach above)?

Reply:
Thanks for checking cookie sizes. That's not good news. My bad I did not check the cookie sizes myself :/ On the other hand, the http://browsercookielimits.squawky.net/ has latest update from 2013 and the most important browsers mentioned there has either no limit for domain or limit of 10234 (IE 8-15). Also the cookie is just a fallback and hopefully won't be needed in most of the cases.

It seems the options are:
- Don't do anything and not implement advanced design. Which means keep client data in the authSessions on the server. This can be done in case that minimal approach will address most of the use-cases and people won't create anymore usability JIRAs :) I fear this won't be the case and better usability may be needed.

- Count with the cookie limit based on the user-agent header. We will know the browser due the ua-parser recently added. In most of the cases, we won't either have limit or have bigger limit like 10234. So we will be able to keep just for example last 3 CLIENT_DATA_* cookies for last 3 shown browser tabs. For the other browser tabs, the redirect to the application will be needed as described in the "minimal" flow.

- Do something else like the POST approach. But I think that POST will be problematic. For example considering all the cases like browser buttons (back/forward/refresh). Currently we remove the POST request from browser history to ensure browser buttons work correctly. I can imagine POST can work just in case that browser is still sending GET requests, but POST requests are sent in the background through ajax. This will require more refactoring of authentication SPI itself to be trully JS based...

- Do something else and still keep the client data on the server rather than in the browser. I can think of something like divide current AuthenticationSession into 2 entities:
-- AuthenticationSession, which will keep just the authentication data
-- ClientAuthenticationSession, which will keep just the client data and reference to the AuthenticationSession. There will be ClientAuthenticationSession for each browser tab, but AuthenticationSession won't be necessarily specific to single browser tab.
However having things on the server doesn't properly solve the problem, that sessions on server can always expire. In case that session will be expired, redirect to the application will be still needed...